### DISCUSSION TITLE
Discriminated Unions Raise "Cannot have a (non-optional) union as a SQLAlchemy field" (Regression?)

### DISCUSSION THREAD

**Discussion by danbreu** [fastapi/sqlmodel#1017] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from sqlmodel import Field, SQLModel
from typing_extensions import Literal


class Cat(SQLModel):
    pet_type: Literal["cat"]
    meows: int


class Dog(SQLModel):
    pet_type: Literal["dog"]
    barks: float


class Lizard(SQLModel):
    pet_type: Literal["reptile", "lizard"]
    scales: bool


class Model(SQLModel, table=True):
    pet: Union[Cat, Dog, Lizard] = Field(..., discriminator="pet_type")
    n: int
```

### Description

Creating a model containing a discriminated union, as shown in the working example in the last comment in #356, raises the exception below.
The only change to the example code is adding `table=True` to `Model`.

```
Traceback (most recent call last):
  File "/data/tool-chain/test.py", line 22, in <module>
    class Model(SQLModel, table=True):
  File "/usr/local/lib/python3.10/dist-packages/sqlmodel/main.py", line 478, in __new__
    col = get_column_from_field(v)
  File "/usr/local/lib/python3.10/dist-packages/sqlmodel/main.py", line 623, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
  File "/usr/local/lib/python3.10/dist-packages/sqlmodel/main.py", line 567, in get_sqlalchemy_type
    type_ = get_type_from_field(field)
  File "/usr/local/lib/python3.10/dist-packages/sqlmodel/_compat.py", line 191, in get_type_from_field
    raise ValueError(
ValueError: Cannot have a (non-optional) union as a SQLAlchemy field
```

### Operating System

Linux

### SQLModel Version

0.0.19

### Python Version

3.10.12

### Additional Context

Pydantic version: 2.8.2
Sqlalchemy version: 2.0.31

**Comment by danbreu** [comment, marked as answer]
This question was based on a misunderstanding of how SQLModel works. There is no way for SQLModel to automatically "collapse" nested models into a top-level table model.
And, as far as I understand, discriminated unions are only available for the Pydantic-based (i.e. non-ORM) part of SQLModel.

**Reply by stuartaxonHO** [reply]
It looks like it should work looking at this, which does seem to have models extending SQLModel

https://github.com/fastapi/sqlmodel/issues/356

I wasn't able to get it working on 0.0.27 though the models were slightly different.
